File: nvidia/src/alert_history.js
#!/usr/bin/env node
/**
 * alert_history.js — separate historian for alert-grade events.
 */
const fs = require('fs');
const path = require('path');

const SOURCE = process.env.ALERT_SOURCE || '/root/wrapper/nvidia/metrics_data/wrapper-events.jsonl';
const OUTPUT = process.env.ALERT_OUTPUT || '/root/wrapper/nvidia/metrics_data/alert-history.jsonl';
const DEDUPE_WINDOW = parseFloat(process.env.ALERT_DEDUPE_WINDOW || '30') * 1000; // ms

// Regular expressions to match alerts based on messages
const RX_429 = /429|rate.limit/i;
const RX_EXHAUST = /exhaust|all.*keys.*failed/i;
const RX_5XX = /5\d\d|upstream.*5\d\d/i;
const RX_UNAVAIL = /unavailable|model.*not.*found/i;
const RX_PACING = /pacing|throttle|enabled_pacing/i;
const RX_DISABLE = /key_disabled|key.*disabled/i;

const dedupe = new Map();

function shouldEmit(kind, model) {
  const key = `${kind}:${model || ''}`;
  const last = dedupe.get(key) || 0;
  const now = Date.now();
  if (now - last < DEDUPE_WINDOW) return false;
  dedupe.set(key, now);
  return true;
}

function classify(ev) {
  const msg = ev.msg || '';
  if (RX_EXHAUST.test(msg)) return { kind: 'exhaustion', severity: 'critical' };
  if (RX_429.test(msg))     return { kind: 'rate_limit', severity: 'warn' };
  if (RX_5XX.test(msg))     return { kind: 'upstream_5xx', severity: 'warn' };
  if (RX_UNAVAIL.test(msg)) return { kind: 'model_unavailable', severity: 'warn' };
  if (RX_PACING.test(msg))   return { kind: 'pacing', severity: 'info' };
  if (RX_DISABLE.test(msg))  return { kind: 'key_disabled', severity: 'warn' };
  return null;
}

function emitAlert(ev, cls) {
  const rec = {
    ts_iso: new Date().toISOString(),
    ts_source: ev.ts,
    kind: cls.kind,
    severity: cls.severity,
    msg: ev.msg,
    model: ev.model,
    key_label: ev.key_label || ev.key,
    attempt: ev.attempt,
    status: ev.status,
    client_ip: ev.client_ip,
    scope: ev.scope,
    in_flight: ev.in_flight,
    scheme: ev.scheme,
    rpm: ev.rpm,
    latency_ms: ev.latency_ms
  };
  // Remove null/undefined fields
  Object.keys(rec).forEach(k => {
    if (rec[k] === undefined || rec[k] === null) delete rec[k];
  });
  return rec;
}

function processLine(line) {
  if (!line.trim()) return;
  try {
    const ev = JSON.parse(line);
    const cls = classify(ev);
    if (!cls) return;
    const rec = emitAlert(ev, cls);
    if (shouldEmit(rec.kind, rec.model)) {
      fs.mkdirSync(path.dirname(OUTPUT), { recursive: true });
      fs.appendFileSync(OUTPUT, JSON.stringify(rec) + '\n');
    }
  } catch (e) {
    // Fail-soft: fallback to basic string-matching if not JSON
    const dummyEv = { msg: line, ts: new Date().toISOString() };
    const cls = classify(dummyEv);
    if (cls && shouldEmit(cls.kind, '')) {
      const rec = { ts_iso: dummyEv.ts, kind: cls.kind, severity: cls.severity, msg: line.trim() };
      fs.mkdirSync(path.dirname(OUTPUT), { recursive: true });
      fs.appendFileSync(OUTPUT, JSON.stringify(rec) + '\n');
    }
  }
}

function modeOnce() {
  if (!fs.existsSync(SOURCE)) {
    console.error(`[alert_history] source missing: ${SOURCE}`);
    process.exit(2);
  }
  const data = fs.readFileSync(SOURCE, 'utf8');
  const lines = data.split('\n');
  let count = 0;
  for (const line of lines) {
    processLine(line);
    count++;
  }
  console.log(`[alert_history] once: processed ${count} lines`);
}

function modeDaemon() {
  let pos = 0;
  try { pos = fs.statSync(SOURCE).size; } catch {}
  console.log(`[alert_history] daemon mode, watching ${SOURCE} (start_pos=${pos})`);

  // Periodically prune dedupe map to prevent unbounded growth
  setInterval(() => {
    const cutoff = Date.now() - DEDUPE_WINDOW * 2;
    for (const [key, ts] of dedupe) {
      if (ts < cutoff) dedupe.delete(key);
    }
  }, Math.max(DEDUPE_WINDOW, 60000));

  setInterval(() => {
    try {
      if (!fs.existsSync(SOURCE)) return;
      const stat = fs.statSync(SOURCE);
      if (stat.size <= pos) return;

      const fd = fs.openSync(SOURCE, 'r');
      const buf = Buffer.alloc(stat.size - pos);
      fs.readSync(fd, buf, 0, buf.length, pos);
      fs.closeSync(fd);
      pos = stat.size;

      const lines = buf.toString('utf8').split('\n');
      for (const line of lines) {
        processLine(line);
      }
    } catch (e) {
      console.error('[alert_history] daemon read error:', e.message);
    }
  }, 500);
}

function modeTop(n = 20) {
  if (!fs.existsSync(OUTPUT)) {
    console.log("(no alert-history yet)");
    return;
  }
  const data = fs.readFileSync(OUTPUT, 'utf8');
  const lines = data.split('\n');
  const seen = new Map();

  for (const line of lines) {
    if (!line.trim()) continue;
    try {
      const rec = JSON.parse(line);
      const k = `${rec.kind}|${rec.model || 'n/a'}|${rec.severity}`;
      if (!seen.has(k)) {
        seen.set(k, { count: 0, last: rec.ts_iso });
      }
      const entry = seen.get(k);
      entry.count++;
      if (rec.ts_iso > entry.last) entry.last = rec.ts_iso;
    } catch {}
  }

  const byCount = Array.from(seen.entries())
    .map(([k, v]) => {
      const [kind, model, severity] = k.split('|');
      return { kind, model, severity, count: v.count, last: v.last };
    })
    .sort((a, b) => b.count - a.count)
    .slice(0, n);

  for (const item of byCount) {
    const sevStr = item.severity.padEnd(8);
    const kindStr = item.kind.padEnd(18);
    const countStr = String(item.count).padStart(5);
    console.log(`${sevStr} | ${kindStr} | ${countStr} alerts | last: ${item.last} | model=${item.model}`);
  }
}

function modeTail() {
  const { spawn } = require('child_process');
  const tailProc = spawn('tail', ['-n', '100', '-f', OUTPUT], { stdio: 'inherit' });
  tailProc.on('error', (err) => console.error('[alert_history] tail error:', err.message));
}

// Command router
const args = process.argv.slice(2);
const modeArgIndex = args.indexOf('--mode');
let mode = 'daemon';
if (modeArgIndex !== -1 && args[modeArgIndex + 1]) {
  mode = args[modeArgIndex + 1];
} else if (args[0] && !args[0].startsWith('-')) {
  mode = args[0];
}

const nArgIndex = args.indexOf('-n');
let n = 20;
if (nArgIndex !== -1 && args[nArgIndex + 1]) {
  n = parseInt(args[nArgIndex + 1], 10) || 20;
}

if (mode === 'once') {
  modeOnce();
} else if (mode === 'top') {
  modeTop(n);
} else if (mode === 'tail') {
  modeTail();
} else {
  modeDaemon();
}

File: nvidia-python/src/alert_history.py
#!/usr/bin/env python3
"""
alert_history.py — separate historian for alert-grade events.
Migrated from alert_history.js — functionally identical.

Reads JSONL events from a source file, classifies them into alert categories,
and writes deduplicated alerts to an output file. Supports once, daemon,
top, and tail modes.
"""

import os
import re
import sys
import time
import json
import asyncio
from pathlib import Path

SOURCE = os.environ.get('ALERT_SOURCE', '/root/wrapper/nvidia/metrics_data/wrapper-events.jsonl')
OUTPUT = os.environ.get('ALERT_OUTPUT', '/root/wrapper/nvidia/metrics_data/alert-history.jsonl')
DEDUPE_WINDOW = float(os.environ.get('ALERT_DEDUPE_WINDOW', '30')) * 1000  # ms

RX_429 = re.compile(r'429|rate.limit', re.IGNORECASE)
RX_EXHAUST = re.compile(r'exhaust|all.*keys.*failed', re.IGNORECASE)
RX_5XX = re.compile(r'5\d\d|upstream.*5\d\d', re.IGNORECASE)
RX_UNAVAIL = re.compile(r'unavailable|model.*not.*found', re.IGNORECASE)
RX_PACING = re.compile(r'pacing|throttle|enabled_pacing', re.IGNORECASE)
RX_DISABLE = re.compile(r'key_disabled|key.*disabled', re.IGNORECASE)

_dedupe: dict = {}


def should_emit(kind: str, model: str) -> bool:
    key = f"{kind}:{model or ''}"
    last = _dedupe.get(key, 0)
    now = time.time() * 1000
    if now - last < DEDUPE_WINDOW:
        return False
    _dedupe[key] = now
    return True


def classify(ev: dict) -> dict:
    msg = ev.get('msg', '')
    if RX_EXHAUST.search(msg):
        return {'kind': 'exhaustion', 'severity': 'critical'}
    if RX_429.search(msg):
        return {'kind': 'rate_limit', 'severity': 'warn'}
    if RX_5XX.search(msg):
        return {'kind': 'upstream_5xx', 'severity': 'warn'}
    if RX_UNAVAIL.search(msg):
        return {'kind': 'model_unavailable', 'severity': 'warn'}
    if RX_PACING.search(msg):
        return {'kind': 'pacing', 'severity': 'info'}
    if RX_DISABLE.search(msg):
        return {'kind': 'key_disabled', 'severity': 'warn'}
    return None


def emit_alert(ev: dict, cls: dict) -> dict:
    rec = {
        'ts_iso': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
        'ts_source': ev.get('ts'),
        'kind': cls['kind'],
        'severity': cls['severity'],
        'msg': ev.get('msg'),
        'model': ev.get('model'),
        'key_label': ev.get('key_label') or ev.get('key'),
        'attempt': ev.get('attempt'),
        'status': ev.get('status'),
        'client_ip': ev.get('client_ip'),
        'scope': ev.get('scope'),
        'in_flight': ev.get('in_flight'),
        'scheme': ev.get('scheme'),
        'rpm': ev.get('rpm'),
        'latency_ms': ev.get('latency_ms'),
    }
    return {k: v for k, v in rec.items() if v is not None}


def process_line(line: str) -> None:
    line = line.strip()
    if not line:
        return
    try:
        ev = json.loads(line)
        cls = classify(ev)
        if not cls:
            return
        rec = emit_alert(ev, cls)
        if should_emit(rec['kind'], rec.get('model', '')):
            Path(OUTPUT).parent.mkdir(parents=True, exist_ok=True)
            with open(OUTPUT, 'a') as f:
                f.write(json.dumps(rec) + '\n')
    except (json.JSONDecodeError, ValueError):
        dummy_ev = {'msg': line, 'ts': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())}
        cls = classify(dummy_ev)
        if cls and should_emit(cls['kind'], ''):
            rec = {'ts_iso': dummy_ev['ts'], 'kind': cls['kind'], 'severity': cls['severity'], 'msg': line}
            Path(OUTPUT).parent.mkdir(parents=True, exist_ok=True)
            with open(OUTPUT, 'a') as f:
                f.write(json.dumps(rec) + '\n')


def mode_once() -> None:
    if not os.path.exists(SOURCE):
        print(f"[alert_history] source missing: {SOURCE}", file=sys.stderr)
        sys.exit(2)
    with open(SOURCE, 'r') as f:
        lines = f.read().split('\n')
    count = 0
    for line in lines:
        process_line(line)
        count += 1
    print(f"[alert_history] once: processed {count} lines")


async def mode_daemon() -> None:
    pos = 0
    try:
        pos = os.path.getsize(SOURCE)
    except OSError:
        pass
    print(f"[alert_history] daemon mode, watching {SOURCE} (start_pos={pos})")

    async def prune_dedupe():
        while True:
            await asyncio.sleep(max(DEDUPE_WINDOW / 1000, 60))
            cutoff = time.time() * 1000 - DEDUPE_WINDOW * 2
            for key in list(_dedupe.keys()):
                if _dedupe[key] < cutoff:
                    del _dedupe[key]

    asyncio.create_task(prune_dedupe())

    while True:
        await asyncio.sleep(0.5)
        try:
            if not os.path.exists(SOURCE):
                continue
            stat = os.stat(SOURCE)
            if stat.st_size <= pos:
                continue
            with open(SOURCE, 'r') as f:
                f.seek(pos)
                buf = f.read(stat.st_size - pos)
            pos = stat.st_size
            for line in buf.split('\n'):
                process_line(line)
        except Exception as e:
            print(f"[alert_history] daemon read error: {e}", file=sys.stderr)


def mode_top(n: int = 20) -> None:
    if not os.path.exists(OUTPUT):
        print("(no alert-history yet)")
        return
    with open(OUTPUT, 'r') as f:
        lines = f.read().split('\n')
    seen = {}
    for line in lines:
        if not line.strip():
            continue
        try:
            rec = json.loads(line)
            k = f"{rec.get('kind', '')}|{rec.get('model', 'n/a')}|{rec.get('severity', '')}"
            if k not in seen:
                seen[k] = {'count': 0, 'last': rec.get('ts_iso', '')}
            entry = seen[k]
            entry['count'] += 1
            if rec.get('ts_iso', '') > entry['last']:
                entry['last'] = rec['ts_iso']
        except (json.JSONDecodeError, ValueError):
            pass

    by_count = sorted(
        [{'kind': k.split('|')[0], 'model': k.split('|')[1], 'severity': k.split('|')[2],
          'count': v['count'], 'last': v['last']} for k, v in seen.items()],
        key=lambda x: x['count'],
        reverse=True,
    )[:n]

    for item in by_count:
        sev_str = item['severity'].ljust(8)
        kind_str = item['kind'].ljust(18)
        count_str = str(item['count']).rjust(5)
        print(f"{sev_str} | {kind_str} | {count_str} alerts | last: {item['last']} | model={item['model']}")


def mode_tail() -> None:
    import subprocess
    try:
        proc = subprocess.Popen(['tail', '-n', '100', '-f', OUTPUT], stdout=sys.stdout, stderr=sys.stderr)
        proc.wait()
    except Exception as e:
        print(f"[alert_history] tail error: {e}", file=sys.stderr)


def main():
    args = sys.argv[1:]
    mode = 'daemon'
    mode_idx = -1
    for i, a in enumerate(args):
        if a == '--mode':
            mode_idx = i
            break
    if mode_idx != -1 and mode_idx + 1 < len(args):
        mode = args[mode_idx + 1]
    elif args and not args[0].startswith('-'):
        mode = args[0]

    n = 20
    n_idx = -1
    for i, a in enumerate(args):
        if a == '-n':
            n_idx = i
            break
    if n_idx != -1 and n_idx + 1 < len(args):
        n = int(args[n_idx + 1]) or 20

    if mode == 'once':
        mode_once()
    elif mode == 'top':
        mode_top(n)
    elif mode == 'tail':
        mode_tail()
    else:
        asyncio.run(mode_daemon())


if __name__ == '__main__':
    main()
